agent/consul/acl_endpoint.go

```diff
@@ -2,239 +2,866 @@ package consul
+	return resetIdx
+}
+
+func (a *ACL) aclPreCheck() error {
+	if !a.srv.ACLsEnabled() {
+		return acl.ErrDisabled
+	}
+
+	if a.srv.UseLegacyACLs() {
+		return fmt.Errorf("The ACL system is currently in legacy mode.")
+	}
+
+	return nil
+}
+
 // Bootstrap is used to perform a one-time ACL bootstrap operation on
 // a cluster to get the first management token.
-func (a *ACL) Bootstrap(args *structs.DCSpecificRequest, reply *structs.ACL) error {
+func (a *ACL) Bootstrap(args *structs.DCSpecificRequest, reply *structs.ACLToken) error {
+	if err := a.aclPreCheck(); err != nil {
+		return err
+	}
 	if done, err := a.srv.forward("ACL.Bootstrap", args, args, reply); done {
 		return err
 	}
 
 	// Verify we are allowed to serve this request
-	if a.srv.config.ACLDatacenter != a.srv.config.Datacenter {
+	if !a.srv.InACLDatacenter() {
 		return acl.ErrDisabled
 	}
 
 	// By doing some pre-checks we can head off later bootstrap attempts
 	// without having to run them through Raft, which should curb abuse.
 	state := a.srv.fsm.State()
-	bs, err := state.ACLGetBootstrap()
+	allowed, resetIdx, err := state.CanBootstrapACLToken()
 	if err != nil {
 		return err
 	}
-	if bs == nil {
-		return structs.ACLBootstrapNotInitializedErr
-	}
-	if !bs.AllowBootstrap {
-		return structs.ACLBootstrapNotAllowedErr
+	var specifiedIndex uint64 = 0
+	if !allowed {
+		// Check if there is a reset index specified
+		specifiedIndex = a.fileBootstrapResetIndex()
+		if specifiedIndex == 0 {
+			return fmt.Errorf("ACL bootstrap already done (reset index: %d)", resetIdx)
+		} else if specifiedIndex != resetIdx {
+			return fmt.Errorf("Invalid bootstrap reset index (specified %d, reset index: %d)", specifiedIndex, resetIdx)
+		}
+
 	}
 
-	// Propose a new token.
-	token, err := uuid.GenerateUUID()
+	accessor, err := lib.GenerateUUID(a.srv.checkTokenUUID)
+	if err != nil {
+		return err
+	}
+	secret, err := lib.GenerateUUID(a.srv.checkTokenUUID)
 	if err != nil {
-		return fmt.Errorf("failed to make random token: %v", err)
+		return err
 	}
 
-	// Attempt a bootstrap.
-	req := structs.ACLRequest{
-		Datacenter: a.srv.config.ACLDatacenter,
-		Op:         structs.ACLBootstrapNow,
-		ACL: structs.ACL{
-			ID:   token,
-			Name: "Bootstrap Token",
-			Type: structs.ACLTypeManagement,
+	req := structs.ACLTokenBootstrapRequest{
+		Token: structs.ACLToken{
+			AccessorID:  accessor,
+			SecretID:    secret,
+			Description: "Bootstrap Token (Global Management)",
+			Policies: []structs.ACLTokenPolicyLink{
+				{
+					ID: structs.ACLPolicyGlobalManagementID,
+				},
+			},
+			CreateTime: time.Now(),
+			Local:      false,
+			// DEPRECATED (ACL-Legacy-Compat) - This is used so that the bootstrap token is still visible via the v1 acl APIs
+			Type: structs.ACLTokenTypeManagement,
 		},
+		ResetIndex: specifiedIndex,
 	}
-	resp, err := a.srv.raftApply(structs.ACLRequestType, &req)
+
+	req.Token.SetHash(true)
```

Comment:
Short Answer: yes if we forget it then it could cause some replication issues but no we shouldn't relegate it to a NewToken function.

The biggest example of why we cant is the token update api. The general usage will be:

1. GET /v1/acl/token/:id
2. Modify a field of the token
3. PUT /v1/acl/token/:id
4. HTTP API will then issue an ACL.TokenUpsert RPC request for the token
5. RPC layer receives the request and applies it in Raft

In that flow the HTTP API cannot know whether the token Hash has been recomputed correctly and the token itself is created by JSON deserialization/mapstructure so there is no way to prevent creating tokens with invalid Hashes here. We could recompute the Hash in the HTTP layer but it wouldn't provide much as it would still be optionally done there.

Instead I thought it better to put this in the RPC layer wherever we are apply token/policy modifications within raft. This happens in 3 functions and recomputing the hash is an unconditional step that happens just before applying in raft.

We probably could compute this in the state store as it should be deterministic and not rely on the raft metadata. However as a general rule I like to not "generate" anything within the state store code as then you have to keep mental track of why generating some things is okay and others is not (like UUIDs)